drizzle-orm/src/mysql-core/query-builders/select.types.ts

```diff
@@ -268,3 +362,98 @@ export type MySqlSelect<
 > = MySqlSelectBase<TTableName, TSelection, TSelectMode, PreparedQueryHKTBase, TNullabilityMap, true, never>;
 
 export type AnyMySqlSelect = MySqlSelectBase<any, any, any, any, any, any, any, any>;
+
+export type MySqlSetOperator<
+	TTableName extends string | undefined = string | undefined,
+	TSelection extends ColumnsSelection = Record<string, any>,
+	TSelectMode extends SelectMode = SelectMode,
+	TPreparedQueryHKT extends PreparedQueryHKTBase = PreparedQueryHKTBase,
+	TNullabilityMap extends Record<string, JoinNullability> = Record<string, JoinNullability>,
+> = MySqlSelectBase<
+	TTableName,
+	TSelection,
+	TSelectMode,
+	TPreparedQueryHKT,
+	TNullabilityMap,
+	true,
+	MySqlSetOperatorExcludedMethods
+>;
+
+export type SetOperatorRightSelect<
+	TValue extends MySqlSetOperatorWithResult<TResult>,
+	TResult extends any[],
+> = TValue extends MySqlSetOperatorInterface<any, any, any, any, any, any, any, infer TValueResult, any>
+	? TValueResult extends Array<infer TValueObj> ? ValidateShape<
+			TValueObj,
+			TResult[number],
+			TypedQueryBuilder<any, TValueResult>
+		>
+	: never
+	: TValue;
+
```

Comment:
Use the `_` property reference instead of inference whenever possible, it's cleaner and was specifically designed for such cases.
```suggestion
> = ValidateShape<
	TValue['_']['result'][number],
	TResult[number],
	TypedQueryBuilder<any, TValue['_']['result']>
>;
```

Reply:
I wish I had thought of that. You have no idea how long I spent with this stupid type. I'll commit in my code to run all the tests again

Reply:
For some reason if I don't narrow the type the way I implemented it, the ValidateShape won't pick it up and won't show the error. I added your suggested change, but the type errors don't show. If you can set it up and check the type-tests let me know. I did however simplified it a little based on your suggestion.

Reply:
I think I know the reason. `TValue` extends `MySqlSetOperatorWithResult<TResult>` and the `TResult` is pulled from the type we actually want to validate against. When we compare `TValue['_']['result'][number]` against `TResult[number]` the compiler is basically comparing the same type against itself.
That's why narrowing _**and inferring**_ is necessary for the ValidateShape to work properly.

Reply:
Makes sense, I might look into the types later to see if they can be further optimized. Type performance is our bottleneck in general, so we have to be careful not to introduce too much type checking overhead.